docs/decisions/0011-producing-to-event-bus-via-settings.rst

```diff
@@ -0,0 +1,106 @@
+11. Enable producing to event bus via settings
+##############################################
+
+Status
+******
+
+**Provisional** 2023-05-23
+
+Context
+*******
+
+Since the initial implementation of the event bus allowed only a single event
+type to be published to a topic, details like topic name, consumer group name,
+and consumer name were configured via code. The current implementation of
+openedx-events does not actually push any events to the underlying
+implementations like `edx-event-bus-kafka`_ and `edx-event-bus-redis`_. The
+event-producing application is expected to create a signal handler (since
+openedx-events subclasses Django signals) to catch the event and push it into
+the event bus. Some examples of the handlers can be found `here`_.
+
+.. _here: https://github.com/openedx/edx-platform/blob/27b8d2f68d5dfaf84755e7d7f8dccc97ce3be509/cms/djangoapps/contentstore/signals/handlers.py#L162-L210
+.. _edx-event-bus-kafka: https://github.com/openedx/event-bus-kafka
+.. _edx-event-bus-redis: https://github.com/openedx/event-bus-redis
+
+This ADR aims to propose a solution for configuring the details, like the topic.
+name, consumer group name, etc. via Django settings as well as pushing events to
+the event bus without requiring the producing application to write additional
+handlers.
+
+
+Decision
+********
+
+Create a generic signal handler to push events to the event bus. This handler
+should be attached to or connected to the signals that are enabled in Django settings.
+of the application. The configuration format will be as shown below:
+
+.. code-block:: python
+
+   OPEN_EDX_EVENTS_CONFIG = {
+       'org.openedx.content_authoring.xblock.deleted.v1': {
+           'topics': 'xblock-status',
```

Comment:
Even if we end up mostly having single-topic events, I have a preference for always requiring a list here. Interested to hear how others feel about that.

Reply:
+1

Reply:
cc @mariajgrimaldi

Reply:
1. If we go down the path of two settings, I think `EXTRA` is a term I've seen for `ADDITIONAL`.
2. We could also consider just overriding by keys like we do for some settings.

Reply:
> Then a deployer could just append to the end of the list, and there could be multiple entries for a single event, with different topics.

@timmc-edx Makes sense. Like the idea of having a list to easily add more topics.

> This still doesn't cover the case of "what if someone wants to exclude a topic for some event?" so it might not be exactly what we want. But I do think this is an issue that's likely to come up, and we should have a story for how are going to manage it.

How about using list of objects for each topic under an event, something like below:

```python
{
    'org.openedx.content_authoring.xblock.published.v1': [
        {'topic': 'content-authoring-xblock-lifecycle', 'event_key_field': 'xblock_info.usage_key', 'enabled': True},
        {'topic': 'xblock-published', 'event_key_field': 'xblock_info.usage_key', 'enabled': False},
    ],
    'org.openedx.content_authoring.xblock.deleted.v1': [
        {'topic': 'content-authoring-xblock-lifecycle', 'event_key_field': 'xblock_info.usage_key', 'enabled': True},
    ],
}
```
This would allow us to set different `event_key_field` for each topic as well as exclude some topics for some events without overriding the full base configuration.

Reply:
@navinkarkera: I'm not clear on how the list can provide partial override without a `BASE` vs `EXTRA` setting. I do understand how we might take original per event type, unless an overridden event type is supplied. 

I'm fine with `event_key_field` being per topic, just in case we ever want something different.

Reply:
>@navinkarkera: I'm not clear on how the list can provide partial override without a BASE vs EXTRA setting. I do understand how we might take original per event type, unless an overridden event type is supplied.

Sorry I wasn't clear before. We will need `BASE` and `EXTRA`. I am proposing the format of the config for them. The `EXTRA` can override and disable or enable a event with same topic name from base config. One more option would be to have something like below which makes merging base and extra simple but the config doesn't look super nice.

```python
{
    'org.openedx.content_authoring.xblock.published.v1': {
        'content-authoring-xblock-lifecycle': {'event_key_field': 'xblock_info.usage_key', 'enabled': True},
        'xblock-published': {'event_key_field': 'xblock_info.usage_key', 'enabled': False},
    },
    'org.openedx.content_authoring.xblock.deleted.v1': {
        'content-authoring-xblock-lifecycle': {'event_key_field': 'xblock_info.usage_key', 'enabled': True},
    },
}
```

Reply:
>  I'm even wondering if BASE vs EXTRA is more complicated than we want, and if we should have a simple setting, and expect it to be 100% overridden and explicit?

Agreed. We could go with the simple approach initially and upgrade if required. Just have a single setting dictionary as below:

```python
{
    'org.openedx.content_authoring.xblock.published.v1': [
        {'topic': 'content-authoring-xblock-lifecycle', 'event_key_field': 'xblock_info.usage_key', 'enabled': True},
        {'topic': 'xblock-published', 'event_key_field': 'xblock_info.usage_key', 'enabled': False},
    ],
    'org.openedx.content_authoring.xblock.deleted.v1': [
        {'topic': 'content-authoring-xblock-lifecycle', 'event_key_field': 'xblock_info.usage_key', 'enabled': True},
    ],
}
```

cc: @timmc-edx

Reply:
@timmc-edx @robrap What do you think about going ahead with the simple approach?

Reply:
It seems like a reasonable place to start to me.

Reply:
Yes, I think starting with that makes sense.

People who want to override can do so carefully by copying the base config and then commenting out (rather than deleting) anything they want to omit, and adding a comment next to anything they add. It's not perfect, but it seems close enough.

It might become an issue once the event bus becomes a mandatory part of Open edX deployments (I think this will happen eventually) because then a mandatory signal/topic pair may be added to a field and deployers might not know to add it to their overrides. Maybe that's not a real problem, though! Our release notes would have to tell people to make the topics in advance, and that would also be an opportunity to tell them to ensure any overrides include the new topic.

Reply:
Thanks for the inputs. @timmc-edx @robrap 

Updated the ADR to reflect the changes in our discussion.